File: rb/test/paranoia_entity_test.rb
# Paranoia entity test

require "minitest/autorun"
require "json"
require_relative "../TruthOrDare_sdk"
require_relative "runner"

class ParanoiaEntityTest < Minitest::Test
  def test_create_instance
    testsdk = TruthOrDareSDK.test(nil, nil)
    ent = testsdk.Paranoia(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = paranoia_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "paranoia." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set TRUTHORDARE_TEST_PARANOIA_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    paranoia_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.paranoia")))
    paranoia_ref01_data = nil
    if paranoia_ref01_data_raw.length > 0
      paranoia_ref01_data = Helpers.to_map(paranoia_ref01_data_raw[0][1])
    end

    # LOAD
    paranoia_ref01_ent = client.Paranoia(nil)
    paranoia_ref01_match_dt0 = {
      "id" => paranoia_ref01_data["id"],
    }
    paranoia_ref01_data_dt0_loaded, err = paranoia_ref01_ent.load(paranoia_ref01_match_dt0, nil)
    assert_nil err
    paranoia_ref01_data_dt0_load_result = Helpers.to_map(paranoia_ref01_data_dt0_loaded)
    assert !paranoia_ref01_data_dt0_load_result.nil?
    assert_equal paranoia_ref01_data_dt0_load_result["id"], paranoia_ref01_data["id"]

  end
end

def paranoia_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "paranoia", "ParanoiaTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = TruthOrDareSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["paranoia01", "paranoia02", "paranoia03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["TRUTHORDARE_TEST_PARANOIA_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "TRUTHORDARE_TEST_PARANOIA_ENTID" => idmap,
    "TRUTHORDARE_TEST_LIVE" => "FALSE",
    "TRUTHORDARE_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["TRUTHORDARE_TEST_PARANOIA_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["TRUTHORDARE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = TruthOrDareSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["TRUTHORDARE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["TRUTHORDARE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: ts/test/entity/paranoia/ParanoiaEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { TruthOrDareSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('ParanoiaEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when TRUTHORDARE_TEST_LIVE=TRUE.
  afterEach(liveDelay('TRUTHORDARE_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = TruthOrDareSDK.test()
    const ent = testsdk.Paranoia()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.TRUTH_OR_DARE_TEST_LIVE
    for (const op of ['load']) {
      if (maybeSkipControl(t, 'entityOp', 'paranoia.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set TRUTH_OR_DARE_TEST_PARANOIA_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let paranoia_ref01_data = Object.values(setup.data.existing.paranoia)[0] as any

    // LOAD
    const paranoia_ref01_ent = client.Paranoia()
    const paranoia_ref01_match_dt0: any = {}
    paranoia_ref01_match_dt0.id = paranoia_ref01_data.id
    const paranoia_ref01_data_dt0 = await paranoia_ref01_ent.load(paranoia_ref01_match_dt0)
    assert(paranoia_ref01_data_dt0.id === paranoia_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/paranoia/ParanoiaTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = TruthOrDareSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['paranoia01','paranoia02','paranoia03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['TRUTH_OR_DARE_TEST_PARANOIA_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'TRUTH_OR_DARE_TEST_PARANOIA_ENTID': idmap,
    'TRUTH_OR_DARE_TEST_LIVE': 'FALSE',
    'TRUTH_OR_DARE_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['TRUTH_OR_DARE_TEST_PARANOIA_ENTID']

  const live = 'TRUE' === env.TRUTH_OR_DARE_TEST_LIVE

  if (live) {
    client = new TruthOrDareSDK(merge([
      {
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.TRUTH_OR_DARE_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  
